File: shared/protocol/protocol-version.ts
export const PROTOCOL_MAJOR = 1;
export const PROTOCOL_MINOR = 0;

export type ProtocolVersion = {
  protocolMajor: number;
  protocolMinor: number;
};

export type ProtocolCompatibility =
  | { compatible: true; negotiatedMinor: number }
  | { compatible: false; reason: string };

export function checkProtocolCompatibility(remote: ProtocolVersion): ProtocolCompatibility {
  if (remote.protocolMajor !== PROTOCOL_MAJOR) {
    return {
      compatible: false,
      reason: `Unsupported protocol major ${remote.protocolMajor}; expected ${PROTOCOL_MAJOR}.`
    };
  }

  return {
    compatible: true,
    negotiatedMinor: Math.min(PROTOCOL_MINOR, remote.protocolMinor)
  };
}

File: android/app/src/main/java/app/prompter/tablet/protocol/ProtocolVersion.kt
package app.prompter.tablet.protocol

object ProtocolVersion {
    const val MAJOR = 1
    const val MINOR = 2

    fun requireCompatible(major: Int, minor: Int): Int {
        require(major == MAJOR) { "Unsupported protocol major $major; expected $MAJOR." }
        return minOf(MINOR, minor)
    }
}
